File: Design_and_analysis_of_algorithms/mergesort.c
#include<stdio.h>
#include<stdlib.h>
int insert=0,merge=0;
void insertsort(int *arr,int n)
{
	int i,key,j;
	for(i=1;i<n;i++)
	{
		key=arr[i];
		j=i-1;
		while(key<arr[j]&&j>=0)
		{
			arr[j+1]=arr[j];
			j--;
			insert++;
		}
		arr[j+1]=key;
		
	}
}

void mergesort(int *arr,int start,int end)
{
	int mid=(start+end)/2;
	int *a,*b,*c;
	int i=0,j=0,k=0;
	if(start>=end)
	{
		return;
	}
	a=(int*)malloc(sizeof(int)*(mid-start+1));
	b=(int*)malloc(sizeof(int)*(end-mid));
	
	c=(int*)malloc(sizeof(int)*(end-start+1));
	
	mergesort(arr,start,mid);
	mergesort(arr,mid+1,end);
	for(i=start;i<=mid;i++)
	{
		a[i-start]=arr[i];
	}
	for(i=mid+1;i<=end;i++)
	{
		b[i-mid-1]=arr[i];
	}
	
	for(k=0,i=0,j=0;k<(end-start+1);k++)
	{
		if(i>mid-start)
		{
			c[k]=b[j];
			j++;
		}
		else if(j>end-mid-1)
		{
			c[k]=a[i];
			i++;
		}
		else if(a[i]<b[j])
		{
			merge++;
			c[k]=a[i];
			i++;
		}
		else if(b[j]<a[i])
		{
			merge++;
			c[k]=b[j];
			j++;
		}
	}
	for(i=start;i<=end;i++)
	{
		arr[i]=c[i-start];
	}
	
	
	
}
int main()
{
	int arr1[]={4,6,2,7,8,2,10};
	int arr2[]={4,6,2,7,8,2,10};
	int i;
	int size=7;
	
	insertsort(arr1,7);
	printf("Insert sort comparisions:%d\n",insert);
	for(i=0;i<size;i++)
	{
		printf("%d\t",arr1[i]);
	}
	printf("\nMerge sort comparisions:%d\n",merge);
	mergesort(arr2,0,6);
	for(i=0;i<size;i++)
	{
		printf("%d\t",arr2[i]);
	}
}
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	

File: Merge_Sort/Merge_Sort.kt
import java.util.*

fun merge(arr: IntArray, left: Int, right: Int, mid: Int)
{
    //Initializing Array
    var arr1Size = mid - left + 1
    var arr2Size = right - mid
    var arr1 = IntArray(arr1Size)
    var arr2 = IntArray(arr2Size)

    //Copy Data of array into arr1 and arr2
    for (i in 0 until arr1Size)
    {
        arr1[i] = arr[left + i]
    }
    for (i in 0 until arr2Size)
    {
        arr2[i] = arr[mid + 1 + i]
    }
    var index1 = 0
    var index2 = 0
    var k: Int = left
    // Merge arr1 and arr2
    while (index1 < arr1Size && index2 < arr2Size)
    {
        if (arr1[index1] <= arr2[index2] )
        {
            arr[k] = arr1[index1]
            index1++
        }
        else
        {
            arr[k] = arr2[index2]
            index2++
        }
        k++
    }
    // Inserting remaining element
    while (index1 < arr1Size)
    {
        arr[k] = arr1[index1]
        index1++
        k++
    }
    while (index2 < arr2Size)
    {
        arr[k] = arr2[index2]
        index2++
        k++
    }
}

fun mergeSort(arr: IntArray, left: Int, right: Int)
{
    if (left < right)
    {
        var mid: Int = left + (right - left) / 2
        mergeSort(arr, left, mid)
        mergeSort(arr, mid + 1, right)
        merge(arr, left, right, mid)
    }
}

// Driver Function for Merge Sort
fun main()
{
    var read = Scanner(System.`in`)
    println("Enter the size of Array")
    var arrSize = read.nextLine().toInt()
    var arr = IntArray(arrSize)
    println("Enter the Elements")
    for (i in 0 until arrSize)
    {
        arr[i] = read.nextLine().toInt()
    }
    mergeSort(arr, 0, arrSize-1)
    println("Sorted Array is:")
    for (i in 0 until arrSize)
    {
        print("${arr[i]} ")
    }
}

/*
OUTPUT:-
Enter the size of Array
 4
Enter the Elements
 2
 5
 3
 1
Sorted Array is:
 1 2 3 5
*/

